spacy/tokens/doc.pyx

```diff
@@ -1650,20 +1666,40 @@ cdef class Doc:
             for span_group in self.spans:
                 data["spans"][span_group] = []
                 for span in self.spans[span_group]:
-                    span_data = {
-                        "start": span.start_char, "end": span.end_char, "label": span.label_, "kb_id": span.kb_id_
-                    }
+                    span_data = {"start": span.start_char, "end": span.end_char, "label": span.label_, "kb_id": span.kb_id_}
                     data["spans"][span_group].append(span_data)
 
         if underscore:
+            user_keys = set()
             data["_"] = {}
+            if self.user_data:
+                data["user_data"] = {"spans":{}, "tokens":{}}
+                for data_key in self.user_data:
+                    if type(data_key) == tuple and len(data_key) >= 4 and data_key[0] == "._.":
+                        attr = data_key[1]
+                        start = data_key[2]
+                        end = data_key[3]
+                        if attr in underscore:
+                            user_keys.add(attr)
+                            value = self.user_data[data_key]
+                            if not srsly.is_json_serializable(value):
+                                raise ValueError(Errors.E107.format(attr=attr, value=repr(value)))
+                            # Check if doc attribute
+                            if start is None:
+                                data["_"][attr] = value
+                            # Check if token attribute
+                            elif end is None:
+                                if attr not in data["user_data"]["tokens"]:
+                                    data["user_data"]["tokens"][attr] = {"token_start": start, "token_end": end, "value":value}
+                            # Else span attribute
+                            else:
+                                if attr not in data["user_data"]["spans"]:
+                                    data["user_data"]["spans"][attr] = {"token_start": start, "token_end": end, "value":value}
```

Comment:
You shouldn't need the `"tokens"` and `"spans"` keys here?

Reply:
I'm also having a naming crisis about `user_data` here again.

Reply:
I added the "tokens" / "spans" keys because they can have custom extensions with the same identifier, so my initial thought was to separate both

Reply:
But you can distinguish them by the `end_token` value just like it's stored internally? Or there's no need for `token_end` for the token spans?

Let me continue my naming crisis, too (with a side of "this looks like prodigy except that `token_end` is inclusive in prodigy" on top).

I wouldn't want to call it this except for `underscore` in the kwarg, but let me think a bit...

In an ideal world we would have:

```
data["_"][attr] = [{"token_start": , "token_end": , "value": }, ]
```

We are not in this ideal world because we are stuck with:

```
data["_"][attr] = doc_ext_value
```

So something that doesn't overlap with other uses of `user_data` but still does something similar:

```
data["underscore_token"][attr] = [{"token_start": , "value": }, ]
data["underscore_span"][attr] = [{"token_start": , "token_end": , "value": }, ]
```

This leaves room to actually add other `user_data` in the future (although I doubt we would).

Reply:
I agree, I've implemented your suggestion and adjusted all tests accordingly